main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionInfoCache.cs

```diff
@@ -23,181 +23,146 @@
+				VersionControlService.NotifyFileStatusChanged (args);
 			}
 		}
 
 		public VersionInfo GetStatus (FilePath localPath)
-		{
-			try {
-				fileLock.EnterReadLock ();
-
-				VersionInfo vi;
-				fileStatus.TryGetValue (localPath, out vi);
-				return vi;
-			} finally {
-				fileLock.ExitReadLock ();
-			}
+		{
+			fileStatus.TryGetValue (localPath, out var vi);
+
+			return vi;
 		}
 
 		public DirectoryStatus GetDirectoryStatus (FilePath localPath)
-		{
-			try {
-				directoryLock.EnterReadLock ();
-
-				DirectoryStatus vis;
-				if (directoryStatus.TryGetValue (localPath.CanonicalPath, out vis))
-					return vis;
-				return null;
-			} finally {
-				directoryLock.ExitReadLock ();
-			}
+		{
+			if (directoryStatus.TryGetValue (localPath.CanonicalPath, out var vis))
+				return vis;
+			return null;
 		}
 
 		public void SetStatus (VersionInfo versionInfo, bool notify = true)
-		{
-			try {
-				fileLock.EnterWriteLock ();
+		{
+			if (!versionInfo.IsInitialized)
+				versionInfo.Init (repo);
 
-				if (!versionInfo.IsInitialized)
-					versionInfo.Init (repo);
-				VersionInfo vi;
-				if (fileStatus.TryGetValue (versionInfo.LocalPath, out vi) && vi.Equals (versionInfo)) {
-					vi.RequiresRefresh = false;
-					return;
-				}
-				fileStatus [versionInfo.LocalPath] = versionInfo;
-			} finally {
-				fileLock.ExitWriteLock ();
-			}
+			if (fileStatus.TryGetValue (versionInfo.LocalPath, out var vi) && vi.Equals (versionInfo)) {
+				vi.RequiresRefresh = false;
+				return;
+			}
+
+			fileStatus.AddOrUpdate (versionInfo.LocalPath, versionInfo, (f, v) => versionInfo);
 
 			if (notify)
 				VersionControlService.NotifyFileStatusChanged (new FileUpdateEventArgs (repo, versionInfo.LocalPath, versionInfo.IsDirectory));
 		}
 
 		public void SetStatus (IEnumerable<VersionInfo> versionInfos)
 		{
-			FileUpdateEventArgs args = null;
-
-			try {
-				fileLock.EnterWriteLock ();
-				foreach (var versionInfo in versionInfos) {
-					if (!versionInfo.IsInitialized)
-						versionInfo.Init (repo);
-					VersionInfo vi;
-					if (fileStatus.TryGetValue (versionInfo.LocalPath, out vi) && vi.Equals (versionInfo)) {
-						vi.RequiresRefresh = false;
-						continue;
-					}
-					fileStatus [versionInfo.LocalPath] = versionInfo;
-					var a = new FileUpdateEventArgs (repo, versionInfo.LocalPath, versionInfo.IsDirectory);
-					if (args == null)
-						args = a;
-					else
-						args.MergeWith (a);
-				}
-			} finally {
-				fileLock.ExitWriteLock ();
-			}
+			FileUpdateEventArgs args = null;
+
+			foreach (var versionInfo in versionInfos) {
+				if (!versionInfo.IsInitialized)
+					versionInfo.Init (repo);
+
+				if (fileStatus.TryGetValue (versionInfo.LocalPath, out var vi) && vi.Equals (versionInfo)) {
+					vi.RequiresRefresh = false;
+					continue;
+				}
+
+				fileStatus.AddOrUpdate (versionInfo.LocalPath, versionInfo, (f, v) => versionInfo);
```

Comment:
```suggestion
			fileStatus[versionInfo.LocalPath] = versionInfo;
```

The setter internally handles concurrent scenarios.